File: flutter_app/lib/features/child/domain/child_experience.dart
/// The complete child-facing contract. It deliberately cannot carry adult notes.
class ChildExperience {
  const ChildExperience({
    required this.sessionId,
    required this.sensory,
    required this.puzzle,
    required this.celebration,
  });

  final String sessionId;
  final SensoryConfiguration sensory;
  final ChildPuzzlePayload puzzle;
  final ChildCelebration celebration;
}

class SensoryConfiguration {
  const SensoryConfiguration({
    required this.reduceMotion,
    required this.soundEnabled,
    required this.hapticsEnabled,
    required this.highContrast,
    required this.themeName,
  });

  final bool reduceMotion;
  final bool soundEnabled;
  final bool hapticsEnabled;
  final bool highContrast;
  final String themeName;
}

sealed class ChildPuzzlePayload {
  const ChildPuzzlePayload({required this.id, required this.seed});

  final String id;
  final int seed;
}

class TimelinePuzzlePayload extends ChildPuzzlePayload {
  const TimelinePuzzlePayload({
    required super.id,
    required super.seed,
    required this.items,
  });

  final List<TimelineItem> items;
}

class TimelineItem {
  const TimelineItem({required this.id, required this.label, required this.order});

  final String id;
  final String label;
  final int order;
}

class ConstellationPuzzlePayload extends ChildPuzzlePayload {
  const ConstellationPuzzlePayload({
    required super.id,
    required super.seed,
    required this.stars,
  });

  final List<ConstellationStar> stars;
}

class ConstellationStar {
  const ConstellationStar({
    required this.id,
    required this.x,
    required this.y,
    required this.isDifferent,
  });

  final String id;
  final double x;
  final double y;
  final bool isDifferent;
}

class ChildCelebration {
  const ChildCelebration({required this.message});

  final String message;
}

File: backend/src/domain/childExperience.ts
/**
 * This module is the complete child-facing contract. It must remain independent
 * of guardian-only field notes and must contain only play, sensory, and warm
 * non-clinical language.
 */
export type SensoryConfiguration = {
  soundEnabled: boolean;
  hapticsEnabled: boolean;
  reducedMotion: boolean;
  theme: 'calm' | 'night' | 'garden' | 'space';
  contrast: 'standard' | 'high';
};

export type CalendarPuzzle = {
  kind: 'calendar-order';
  prompt: string;
  cards: Array<{ id: string; label: string }>;
};

export type ConstellationPuzzle = {
  kind: 'constellation-anomaly';
  prompt: string;
  stars: Array<{ id: string; x: number; y: number; group: string }>;
  choices: Array<{ id: string; label: string }>;
};

export type ChildExperience = {
  sessionId: string;
  layer: number;
  track: 'calendar-genius' | 'constellation-mapper';
  sensory: SensoryConfiguration;
  puzzle: CalendarPuzzle | ConstellationPuzzle;
  celebration: string;
  canSkip: true;
  canPause: true;
};

const prohibitedChildTerms = [
  'diagnos', 'autis', 'disorder', 'condition', 'career', 'job', 'employ',
  'industry', 'salary', 'income', 'employer', 'assessment', 'screening',
];

export function assertSafeChildExperience(value: ChildExperience): ChildExperience {
  const serialized = JSON.stringify(value).toLowerCase();
  const forbidden = prohibitedChildTerms.find((term) => serialized.includes(term));
  if (forbidden) throw new Error(`Child experience contains prohibited content: ${forbidden}`);
  return value;
}
